feat: add/remove reaction

hannojg opened the pull request:
<!-- If necessary, assign reviewers that know the area or changes well. Feel free to tag any additional reviewers you see fit. -->

### Details
<!-- Explanation of the change or anything fishy that is going on -->

Implementation for adding and removing reactions, according to:

 - https://github.com/Expensify/App/issues/15125

**Known issues:**

- When sending reactions very rapidly it can happen that a pusher event with stale information overwrites the local state, and thus causing buggy behaviour.  This is a known issue with the current architecture. See this video for an example of the bug: https://user-images.githubusercontent.com/16821682/222969430-873513c0-63d5-4a84-95f5-72a13377762e.mov

- When editing a message the reactions will get removed. Here pusher sends an event where the `reactions` property in the message object is missing. This needs to be fixed elsewhere I believe.

**(Done) TODO:**

- [x] https://github.com/Expensify/App/pull/15210/commits/8bc100bd259e52452a1593037406b69de03e0380 Fix issue where components don't receive update to re-render when onyx data changes
- [x] 1cb6d6070e27f22f4cdeb0e5bd552a7e9c7ddbee Fix regression with mini context menu
- [x] https://github.com/Expensify/App/pull/15210/commits/55834f685b0c34644010a3f8176e9609627d1a72 Fix bug where removing a reaction leaves an artefact
  - https://user-images.githubusercontent.com/16821682/222957501-5941c188-4a35-4d05-83ad-7c06fc7c7176.mov
- [x] https://github.com/Expensify/App/pull/15210/commits/e9ba4f526f4599ff29f0a6edcd28817e21204c2b Fix bug:
  - User A add reaction
  - User B react as well
  - User B try to remove reaction -> Doesn't work, instead of removing it tries to add the reaction again





### Fixed Issues
<!---
1. Please replace GH_LINK with a URL link to the GitHub issue this Pull Request is fixing.
2. Please replace PROPOSAL: GH_LINK_ISSUE(COMMENT) with a URL link to your GitHub comment, which contains the approved proposal (i.e. the proposal that was approved by Expensify).

Do NOT add the special GH keywords like `fixed` etc, we have our own process of managing the flow.
It MUST be an entire link to the github issue and your comment proposal ; otherwise, the linking will not work as expected.

Make sure this section looks similar to this (you can link multiple issues using the same formatting, just add a new line):

$ https://github.com/Expensify/App/issues/<number-of-the-issue>
$ https://github.com/Expensify/App/issues/<number-of-the-issue(comment)>

Do NOT only link the issue number like this: $ #<number-of-the-issue>
--->
$ https://github.com/Expensify/App/issues/15125
PROPOSAL: GH_LINK_ISSUE(COMMENT)


### Tests
<!---
Add a numbered list of manual tests you performed that validates your changes work on all platforms, and that there are no regressions present.
Add any additional test steps if test steps are unique to a particular platform.
Manual test steps should be written so that your reviewer can repeat and verify one or more expected outcomes in the development environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->

#### Quick reactions are working

**Test 1: Web specific, hover menu:**

- Login with user A
- Open a chat with user B
- Send a message
- Hover over the send message with the mouse
- Expect the context menu to show in the top right corner, and it shows the four quick reaction emojis as well as a icon button to open the emoji picker
- When pressing on the "👍" reaction, the "👍" gets added as reaction to the message
- When pressing the "👍" in the menu again the reaction gets removed

**Test 2: Quick reactions from context menu**

- Login with user A
- Open a chat with user B
- Send a message
- When right pressing (on web/desktop), or long pressing (mobile) the context menu opens
- When pressing on the "👍" reaction, the "👍" gets added as reaction to the message
- The context menu closes after pressing the reaction
- Open the context menu again and press the "👍" in the menu again -> the reaction gets removed

#### The emoji picker is working

**Test 1: Web specific, hover menu:**

- Login with user A
- Open a chat with user B
- Send a message
- Hover over the send message with the mouse
- Expect the context menu to show in the top right corner, and it shows the four quick reaction emojis as well as a icon button to open the emoji picker
- Press the icon button with the emoji to open the emoji picker 
- Expect the hover menu to stay open while the emoji picker opens next to the icon button
- When selecting an emoji it gets added to the message as reaction
- Upon selecting an emoji the emoji picker closes
- When opening the emoji picker again and pressing the same emoji again it gets removed from the message

**Test 2: Emoji picker from context menu**

- Login with user A
- Open a chat with user B
- Send a message
- When right pressing (on web/desktop), or long pressing (mobile) the context menu opens
- Press the icon button with the emoji to open the emoji picker 
- Expect the context menu to dismiss and instead the emoji picker shows
- When selecting an emoji it gets added to the message as reaction
- Upon selecting an emoji the emoji picker closes
- When opening the emoji picker again and pressing the same emoji again it gets removed from the message

**Test 3: Emoji picker from reactions**
- Login with user A
- Open a chat with user B
- Send a message
- Add a reaction by either using the picker or quick reactions, doesn't matter
- Expect that next to the reaction that now shows underneath the message a "add emoji" icon button appeared
- When pressing it expect the emoji picker to open
- When selecting an emoji it gets added to the message as reaction
- Upon selecting an emoji the emoji picker closes
- When opening the emoji picker again and pressing the same emoji again it gets removed from the message

#### 📩 Sending & receiving works
- Login with user A on device A
- Login with user B on device B
- User A sends a chat message to user B
- User B reacts with a :+1:  (👍) to the message of user A
- Both user A and user B can see the reaction on the message user A sent
- For user B the reaction underneath the message is highlighted (so it's obvious to user B that he/she added the reaction)
- User A now reacts as well with a :+1:
- For both users the reaction counter increments to 2
- Both users now remove their reaction
- For both users the reaction is gone
- User A now adds another emoji reaction, which isn’t the :+1: (e.g. :blush: 😊)
- After that User B now adds again a :+1: reaction
- We confirm that for both users the :+1: reaction is after the :blush: reaction user A sent (chronological order) 

#### Adding the same reaction twice with different skin tone 
- Login with user A
- Open a chat with user B
- Send a message
- Add a "👍" as reaction
- Now open the emoji picker for adding a reaction
- Change the default skin color to something else than the yellow tone
- Add the "👍" emoji again but this time with the different skin tone
- Verify that the reaction bubble underneath the message now shows both emoji skin tones and the counter says "2". Also make sure that the emoji appear in the correct order (e.g. first the yellow one, then the one with the skin tone that you picked)

### Offline tests
<!---
Add any relevant steps that validate your changes work as expected in a variety of network states e.g. "offline", "spotty connection", "slow internet", etc. Manual test steps should be written so that your reviewer and QA testers can repeat and verify one or more expected outcomes. If you are unsure how the behavior should work ask for advice in the `#expensify-open-source` Slack channel.
--->

- Login with user A on device A
- Login with user B on device B
- User A sends a chat message to user B
- After user B receives the message 
- Put the device of user B offline
- User B reacts with a :+1: to the message of user A
- User B sees the reaction immediately, while user A doesn't yet
- User B goes online again
- User A receives the emoji reaction User B sent


### QA Steps
<!---
Add a numbered list of manual tests that can be performed by our QA engineers on the staging environment to validate that your changes work on all platforms, and that there are no regressions present.
Add any additional QA steps if test steps are unique to a particular platform.
Manual test steps should be written so that the QA engineer can repeat and verify one or more expected outcomes in the staging environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->

See testing steps above 

### PR Author Checklist
<!--
This is a checklist for PR authors. Please make sure to complete all tasks and check them off once you do, or else your PR will not be merged!
-->

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is correct English and approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

https://user-images.githubusercontent.com/16821682/220105639-1dcc4ada-7ba7-4746-bb59-466a29089dd4.mov

</details>

<details>
<summary>Mobile Web - Chrome</summary>

https://user-images.githubusercontent.com/16821682/219958885-c930ef2e-bbf1-41c0-8c2e-f3f79d5522b3.mov

</details>

<details>
<summary>Mobile Web - Safari</summary>

https://user-images.githubusercontent.com/16821682/219958680-0dea93a4-e35c-410a-a19a-cc8be6131d21.mp4


</details>

<details>
<summary>Desktop</summary>

https://user-images.githubusercontent.com/16821682/220105788-3b550103-20dd-4585-8cf7-888ab92f3d65.mov

</details>

<details>
<summary>iOS</summary>

https://user-images.githubusercontent.com/16821682/219958760-b6890ee9-1313-420a-bd14-18125e24eba4.mp4


</details>

<details>
<summary>Android</summary>

https://user-images.githubusercontent.com/16821682/219959288-d3302dc4-b6e1-47d4-9d7f-2343171bd694.mov

</details>

hannojg committed:
wip: render quick reactions in context menu

hannojg committed:
wip: adjustments for web

hannojg committed:
add actions for adding and removing reactions

hannojg committed:
wip: add reaction when using quick reactions

hannojg committed:
render emoji reactions underneath a message

hannojg committed:
TEMP: REMOVE ME once accountID on personal details is live

hannojg committed:
change from `login` to `accountID`

hannojg committed:
wip: adding / removing

hannojg committed:
change: when removing reactions, remove the whole emoji object

hannojg committed:
cleanup code

hannojg committed:
show AddReactionBubble in the row of reactions as well

hannojg committed:
implemented toggling reactions when selecting the same again

hannojg committed:
close context menu when open emoji picker or selected emoji

hannojg committed:
fix emoji picker being wrong positioned

hannojg committed:
wip: mini menu

hannojg committed:
wip: mini menu adding add reaction icon

hannojg committed:
fix: mini menu open picker wrong positioned + was closing

hannojg committed:
Merge branch 'main' of github.com:margelo/expensify-app-fork into hanno/feat-add-remove-reactions

hannojg committed:
fix: open emoji picker on mobile

hannojg committed:
clean up

hannojg committed:
rename and document functions

hannojg committed:
pick the preferred skin tone

hannojg committed:
updated JSDocs

hannojg committed:
updated JSDocs

hannojg committed:
updated JSDocs

hannojg committed:
remove todos

hannojg committed:
apply suggestions from design doc

hannojg committed:
apply suggestions from design doc

hannojg commented:
CLA Assistant GitHub actions seems to be broken

hannojg commented:
Oh, there is still one bug, hold on …

hannojg committed:
Merge branch 'main' of github.com:margelo/expensify-app-fork into hanno/feat-add-remove-reactions

hannojg committed:
fix tests

stitesExpensify commented on src/components/Reactions/EmojiReactionBubble.js:
```diff
@@ -0,0 +1,94 @@
+    /**
+     * The account ids of the users who reacted.
+     */
+    senderIDs: PropTypes.arrayOf(PropTypes.oneOfType([PropTypes.string, PropTypes.number])),
```
```suggestion
    reactionUsers: PropTypes.arrayOf(PropTypes.oneOfType([PropTypes.string, PropTypes.number])),
```

stitesExpensify commented on src/components/Reactions/MiniQuickEmojiReactions.js:
```diff
@@ -0,0 +1,83 @@
+ * context menu which we just show on web, when hovering
+ * a message.
+ * @param {Props} props
+ * @returns {JSX.Element}
```
Same thing here, I don't think we need the `param` and `returns`

stitesExpensify commented on src/components/Reactions/MiniQuickEmojiReactions.js:
```diff
@@ -0,0 +1,83 @@
+                >
+                    <Text style={[
+                        styles.emojiReactionText,
+                        StyleUtils.getEmojiReactionTextStyle(1.3),
```
Let's move this to a variable to make it more clear what it is for

stitesExpensify reviewed (commented):
Looking good! Just a couple of quick comments I noticed while reading through

hannojg committed:
fix: pas emoji object directly

hannojg committed:
refactor: pass ref optional as arg

hannojg committed:
fix: pass anchor position for emoji picker as well to fix issue with positioning

hannojg commented on src/components/BaseMiniContextMenuItem.js:
```diff
@@ -0,0 +1,77 @@
+/**
+ * Component that renders a mini context menu item with a
+ * pressable. Also renders a tooltip when hovering the item.
+ * @param {Object} props
```
If I don't, eslint complaints that its not a proper JSDoc comment:

![Screenshot 2023-02-20 at 13 23 25](https://user-images.githubusercontent.com/16821682/220106358-8c3c105d-8d50-4ea7-9947-2bf1baedf2bb.png)

stitesExpensify replied:
Oh okay haha nvm

hannojg committed:
apply suggestions from code review

stitesExpensify commented:
@0xmiroslav if you could give this an initial review today that would be much appreciated!

deetergp reviewed (dismissed):
Super exciting stuff!

stitesExpensify commented:
My apologies if I wasn't clear before. We want to review now so that Hanno can make any necessary changes, and then this can immediately be merged when it goes off of hold

stitesExpensify commented:
@0xmiroslav have you had a chance to take a look at this yet?

stitesExpensify commented on src/libs/actions/Report.js:
```diff
@@ -1176,6 +1182,154 @@ function clearIOUError(reportID) {
+    // const parameters = {
+    //     reportID,
+    //     skinTone,
+    //     reaction: emoji.name,
```
This param is called `emojiCode`

stitesExpensify commented on src/libs/actions/Report.js:
```diff
@@ -1176,6 +1182,154 @@ function clearIOUError(reportID) {
+    //     sequenceNumber: originalReportAction.sequenceNumber,
+    //     reportActionID: originalReportAction.reportActionID,
+    // };
+    // API.write('AddReaction', parameters, {optimisticData});
```
the name of this is endpoint is `AddEmojiReaction`

stitesExpensify commented on src/libs/actions/Report.js:
```diff
@@ -1176,6 +1182,154 @@ function clearIOUError(reportID) {
+    //     reportActionID: originalReportAction.reportActionID,
+    //     reaction: emojiCode,
+    // };
+    // API.write('RemoveReaction', parameters, {optimisticData});
```
`RemoveEmojiReaction`

stitesExpensify commented on src/libs/actions/Report.js:
```diff
@@ -1176,6 +1182,154 @@ function clearIOUError(reportID) {
+    //     reportID,
+    //     sequenceNumber: originalReportAction.sequenceNumber,
+    //     reportActionID: originalReportAction.reportActionID,
+    //     reaction: emojiCode,
```
`emojiCode:`

stitesExpensify reviewed (commented):
Let's update the commands so that this is calling the right thing when the API goes live (Should be on staging today)

hannojg committed:
Merge branch 'main' of github.com:margelo/expensify-app-fork into hanno/feat-add-remove-reactions

hannojg committed:
enable API calls

hannojg committed:
Merge branch 'main' of github.com:margelo/expensify-app-fork into hanno/feat-add-remove-reactions

hannojg commented:
Hey, thx for the review!

I just added the actual API calls, however I am experiencing a bug where Onyx data changes but the connected components do not rerender (if I move the cursor away from the message, thus changing the hovering state the component gets re-rendered with the latest data):

https://user-images.githubusercontent.com/16821682/222920865-4a77888c-08c5-47df-b393-09422934bf48.mov

Not sure if anyone has seen something like that before with onyx? 
Will fix that first and then fix the bugs as named in the review.

hannojg committed:
fix issue where reactions won't re-render

hannojg committed:
temp fix toggling reactions is broken

hannojg commented:
@stitesExpensify When getting the `preferredSkinTone` from onyx its sometimes still `"default"`. I thought we changed that to `-1`? Can you please check what's going on there? I suspect that we didn't change the initial value somewhere?

https://github.com/Expensify/App/pull/15210/commits/e9ba4f526f4599ff29f0a6edcd28817e21204c2b

hannojg committed:
fix removing reactions causing artefacts

hannojg committed:
fix regression where mini menu items weren't displaying correctly

hannojg committed:
fix: don't remove reactions when editing a message

Note: This only works locally. Pusher currently will send an event after the edit that doesn't carry any reactions and as such will overwrite the data

stitesExpensify commented:
> When getting the preferredSkinTone from onyx its sometimes still "default"

Some users probably still have "default" saved in the DB, and it won't be changed until we either:
1. Do a data migration to change that
or
2. The user manually changes their skin tone

For now I think we should just accept both

stitesExpensify commented on src/libs/actions/Report.js:
```diff
@@ -1186,6 +1201,154 @@ function clearIOUError(reportID) {
+    }
+
+    const hasCurrentUserReacted = hasAccountIDReacted(currentUserAccountID, reactionObject.users, skinTone);
+    if (hasCurrentUserReacted) {
```
Let's make this inline instead of creating a var that's only used once

stitesExpensify commented on src/libs/actions/Report.js:
```diff
@@ -1186,6 +1201,154 @@ function clearIOUError(reportID) {
+        }),
+
+        // Remove any null reactions
+        r => r != null,
```
```suggestion
        reportObject => reportObject != null,
```

stitesExpensify reviewed (commented):
Looking great! Just some minor comments!

hannojg committed:
expose hasAccountIDReacted

hannojg committed:
apply code suggestion

hannojg committed:
Update src/libs/actions/Report.js

Co-authored-by: Brandon Stites <42391420+stitesExpensify@users.noreply.github.com>

hannojg committed:
apply code suggestion

hannojg committed:
apply code suggestion

hannojg committed:
apply code suggestion

hannojg committed:
apply code suggestion

hannojg committed:
remove unnecessary requestLayoutAnimation

hannojg committed:
added translations

stitesExpensify commented on src/languages/es.js:
```diff
@@ -224,6 +224,8 @@ export default {
         deleteComment: 'Eliminar comentario',
         deleteConfirmation: '¿Estás seguro de que quieres eliminar este comentario?',
+        addEmojiReaction: 'Añadir una reacción emoji',
+        addReactionTooltip: 'Añadir una reacción...',
```
Hm this looks different than the other one. I think this is just 3 `.` instead of the `...` symbol? Let's update this one to match

stitesExpensify reviewed (commented):
Looking great! 1 more small change and I think we're good to go code wise! Testing everything now

stitesExpensify commented:
@0xmiroslav if you could give this a final review that would be much appreciated!

hannojg committed:
match text

hannojg commented:
Thanks for the early feedback, will incorporate your suggestions now!

hannojg committed:
reflect preferred skin tone in the rest of the UI

hannojg commented:
> can we match selected skin tone everywhere?

added

hannojg commented:
2. Reproduction of the bug


https://user-images.githubusercontent.com/16821682/223191190-16d06b70-f636-4d7c-be65-ea08f3356c43.mov

hannojg committed:
fixed bug where mini context menu wouldn't close

hannojg commented:
2. Bug solved

hannojg committed:
fix: make emojis in quick reaction unselectable

hannojg committed:
fix: immediately remove reaction bar when no reactions are present anymore

hannojg commented:
@0xmiroslav addressed all of your suggestions

hannojg committed:
fix tests

stitesExpensify commented:
@0xmiroslav if we could get another final review that would be great!

hannojg committed:
Merge branch 'main' of github.com:margelo/expensify-app-fork into hanno/feat-add-remove-reactions

stitesExpensify commented:
@0xmiroslav can you also confirm testing on iOS for me? My environment isn't working at the moment

hannojg committed:
fix warning

hannojg committed:
suppress eslint warning

hannojg commented:
@0xmiroslav 

> When double click reaction, final result should be original reaction. For now, it works as if clicked 1 time. Or is this expected?

Sorry what do you mean by that? Do you maybe have a video recording of the behaviour you are describing?

hannojg committed:
remove unused default props

hannojg commented:
@0xmiroslav oh that's a known bug, related to how we receive pusher events.
It has been discussed here: https://expensify.slack.com/archives/C01GTK53T8Q/p1678029220184279

and we agreed on fixing that not in this PR.

hannojg committed:
remove focusable prop as not needed on web

hannojg commented:
Fixed

stitesExpensify reviewed (dismissed):
Looks great!

hannojg committed:
add tooltip

hannojg commented:
The small emoji bubbles underneath the message receive a custom tooltip, as it can be seen here: 

- https://github.com/Expensify/App/pull/15671

I added the tooltip to the context menu emoji reactions

hannojg commented on src/components/Reactions/QuickEmojiReactions/BaseQuickEmojiReactions.js:
```diff
@@ -39,15 +40,17 @@ const propTypes = {
-            />
+            <Tooltip text={`:${emoji.name}:`}>
+                <EmojiReactionBubble
+                    key={emoji.name}
```
man you are the best, thx for spotting those issues <3

hannojg replied:
fixed

hannojg committed:
fix key issue

stitesExpensify commented:
Sounds good. If you could prioritize this in the morning it would be much appreciated as we're trying to get this merged ASAP!

hannojg committed:
fix focusable prop

hannojg commented on src/components/Tooltip/index.js:
```diff
@@ -149,9 +149,7 @@ class Tooltip extends PureComponent {
                 ref={el => this.wrapperView = el}
-                style={this.props.containerStyles}
                 onBlur={this.hideTooltip}
-                focusable
```
Okay, I turned that to a prop, which is default `true`. The perusable component automatically sets focusable to true. So when rendering the reaction bubbles for the context menu we set focusable to false, as the children will handle the focus on their own

hannojg committed:
rename to openContextMenu and shouldKeepOpen

hannojg committed:
Revert "TEMP: REMOVE ME once accountID on personal details is live"

This reverts commit f0926c1d45961ff0c6204f83da5b5a1d4a541fdf.

hannojg commented on src/components/withCurrentUserPersonalDetails.js:
```diff
@@ -22,7 +22,10 @@ export default function (WrappedComponent) {
                 ref={props.forwardedRef}
-                currentUserPersonalDetails={props.personalDetails[currentUserEmail]}
+                currentUserPersonalDetails={{
+                    ...props.session,
```
oh lord, good catch, that was a temp workaround while some backend functionality wasn't there yet. Reverted that change

hannojg committed:
add container styles back

hannojg commented on src/components/Tooltip/index.js:
```diff
@@ -149,9 +149,7 @@ class Tooltip extends PureComponent {
         let child = (
             <View
                 ref={el => this.wrapperView = el}
-                style={this.props.containerStyles}
```
your reviews are the best, well catched. Probably removed that during testing. Added it back

hannojg committed:
Merge branch 'main' of github.com:margelo/expensify-app-fork into hanno/feat-add-remove-reactions

hannojg commented:
The IOU issue is very likely a backend issue. As this PR has urgency for demoing purposes on expansiconx we may skip reactions for IOU messages in the frontend for now, until the backend is fixed properly.

hannojg committed:
skip reactions for IOU messages for now

due to backend issues

stitesExpensify commented:
Great! Looks awesome to me, what about you @0xmiroslav

stitesExpensify reviewed (dismissed):
Looks great!

stitesExpensify commented:
@0xmiroslav we had an internal conversation and decided that because the edit message bug is a backend problem, we are okay with releasing this feature without fixing it. Apologies for not letting you know here on the issue

stitesExpensify commented:
> this issue and IOU message issue are temporary and should be fixed after ECX right?

That is correct, they will be fixed either after or during ECX 😄

stitesExpensify commented:
Nice catch! So that's (half) intentional. You can have multiple of the same reaction if there _are_ different skin tone options, but you shouldn't be able to have multiple of the same reaction if there _aren't_ different skin tone options. 

@hannojg seems like we need to update this logic to include checking if the emoji actually supports skin tones or not 

https://github.com/Expensify/App/pull/15210/files#diff-8afe5b71ee0436c21364148c86dadd640f2bff3e3d916addbb1f1f6f7e5b6a43R1234-R1245

hannojg committed:
fix focus on AddReactionBubble

hannojg commented:
Hey @0xmiroslav & @stitesExpensify in regards to this:

https://github.com/Expensify/App/pull/15210#issuecomment-1459429273

Its working as agreed in the design doc. As in slack you can add the same emoji multiple times to a reaction with different skin tones. Clicking on the group of reactions will remove the whole group (see this video from slack):


https://user-images.githubusercontent.com/16821682/223688978-7e672bfd-f58a-4050-9ef5-b1e22664403b.mov

@stitesExpensify I think the rest of the video is also correct. The message has the following reactions added by the other participant in the chat: "👍🏽", "😄". The user then adds his own reaction to it, and this is working as expected as from what i can see.

(Correct me if i am wrong pls)

hannojg committed:
fix safari selection bug

hannojg commented:
Oh you are absolutely right, will fix it 😊

hannojg committed:
fix adding multiple times the same reaction

hannojg commented:
@0xmiroslav Okay, pushed all fixes. Can you please test again?

hannojg commented:
@0xmiroslav to me that seems like the buggy behaviour with the API where pusher events overwrite local state with stale data

0xmiroslav reviewed (dismissed):
Follow-up issues:
- support reaction on edit message
- support reaction on IOU message
- last reaction reverted when react multiple times quickly

stitesExpensify commented on src/libs/actions/Report.js:
```diff
@@ -1351,12 +1351,13 @@ function removeEmojiReaction(reportID, originalReportAction, emoji) {
+function toggleEmojiReaction(reportID, reportAction, emoji, paramSkinTone = preferredSkinTone) {
     const message = reportAction.message[0];
     const reactionObject = message.reactions && _.find(message.reactions, reaction => reaction.emoji === emoji.name);
+    const skinTone = emoji.types == null ? null : paramSkinTone; // only use skin tone if emoji supports it
```
```suggestion
    const skinTone = emoji.types === null ? null : paramSkinTone; // only use skin tone if emoji supports it
```

stitesExpensify reviewed (commented):
So close, last change!

deetergp reviewed (dismissed):
Thanks for the changes and additional tests!

hannojg committed:
check specifically for undefined

stitesExpensify reviewed (approved):
LGTM and tests well 🎉  LET'S DO THIS!

stitesExpensify commented:
Waiting for verify_podfile....

stitesExpensify commented on src/pages/home/report/ContextMenu/BaseReportActionContextMenu.js:
```diff
@@ -55,28 +62,47 @@ class BaseReportActionContextMenu extends React.Component {
+                    };
+
+                    if (contextAction.renderContent) {
+                        // make sure that renderContent isn't mixed with unsupported props
+                        if (__DEV__ && (contextAction.text != null || contextAction.icon != null)) {
+                            throw new Error('Dev error: renderContent() and text/icon cannot be used together.');
+                        }
```
Honestly I think this is a good warning to have, I wouldn't mind keeping it

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/4370735956) to staging by https://github.com/stitesExpensify in version: 1.2.81-0 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅

puneetlath commented:
Friendly FYI: this PR created a very small regression in the copy: https://github.com/Expensify/App/issues/15867. Should've been `Edit reaction` instead of `Edit Reaction` in order to be consistent with how we do it everywhere else in the app.

stitesExpensify commented:
Ach dangit. Thanks for letting me know

sobitneupane commented:
This PR created a regression: https://github.com/Expensify/App/issues/16384. On navigation to Add reaction bubble using keyboard, the emoji picker appears in top left corner.

sobitneupane commented:
With [this PR](https://github.com/Expensify/App/issues/15815), we have updated the size of emoji in reaction of a message to 28px with a font size of 15px, in mini context menu to 20px with a font size of 13px and in context menu to 34px with a font size of 17px.

Those changes might have been missed or not considered in this PR.

rushatgabhane reviewed (commented):
Small bug fixed: [Clicking on same emoji reaction changes skin tone instead of removing](https://github.com/Expensify/App/issues/46387#top)
